Typescriptify - Conversion to TypeScript.

BurtHarris opened the pull request:
Readme still needs to be updated.

- BREAKING CHANGE: Deprecated decorators removed: `@debounce`, `@throttle`, `@mixin`, and `@memoize`, and associated tests
- All source files now in TypeScript, tests remain in ECMAScript  
- Includes a fast, gulp-based build using [gulp-tsb](https://www.npmjs.com/package/gulp-tsb)    
- Includes [eslint ](https://www.npmjs.com/package/eslint)checking on all .js files, no tslint yet, but soon

The tests remain in ECMAScript, so that they can be run for both Babel and TypeScript generated versions.   Some of the test cases aren't working in TypeScript yet, and _may_ require a typescript upgrade to support them.   But __all__ the remaining tests continue to work when called from Babel processed sources.

Tests now require the `core-decorators` package by name, so they are like any dependent code.   This is supported at test-time by creating a symlink in ./test/node_modules/core-decorators pointing to the project root directory.

Commit:
added package-lock.json generated by npm@5

This file is supposed to allow checking that exactly the intended versions of
dependencies are being used.

Commit:
feat: remove deprecated decorators

Removes the @debounce, @throttle, @mixin, and @memoize decorators which have been deprecated for some time.

The corresponding test cases have been removed as well, bringing the current number of tests expected to 56.

BREAKING CHANGE: but we may not yet increment the major version!  SemVer is currently not applicable to this package due to major version of 0.

Commit:
further work on removing obsolete decorators

Commit:
Update build scripts to use TypeScript exclusively

Commit:
Convert the easy functions to TypeScript

Some edits needed, but nothing major.   Left the hard ones as .js files.

Commit:
Build src with gulp

This includes changin the output directories to cjs and esm.

Commit:
Build tests w/ Babel at build time

They build into directory testBabel.  With this change, we no longer need to use mocha switchs `--compilers js:babel-core/register` or `--require babel-polyfill`

Commit:
Add gulp build of tests for Typescript

Switched default output directory back to `lib`

Switch clean to use del rather than rimraf

Update VSCode launch & tasks config

Fixup top-level tests for TypeScript

Commit:
Fix gulpfile for typescript test sourcemaps

setup VSCode launch to test typescript

Commit:
Push test sources into test/src

Commit:
Revert "Push test sources into test/src"

This reverts commit 4ae646f786fb6c88370e35a05799062b20ed651f.

Commit:
continued gulp/TypeScript conversion

Typescript buidls with gulp-tsb for speed
Integration with babel, eslint and mocha

Commit:
Enable eslint in build

There are a few warnings left.

Commit:
Wrapup making all tests run under both Babel and TypeScript

Some of the TypeScript runs are still failing...

BurtHarris commented on .gitignore:
```diff
@@ -1,5 +1,7 @@
 /es/
-/node_modules/
-/built/
+/esm/
```
I went ahead and changed the `esm` directory to `esm` to help clarify it's purpose.   @jayphelps thanks for describing this for me.

BurtHarris commented on .gitignore:
```diff
@@ -1,5 +1,7 @@
-/built/
+/esm/
+/test/babel/
+/test/typescript/
```
Tests, from both transpilers, now build into these subdirectories so that they can be compared.

BurtHarris commented on index.d.ts:
```diff
@@ -1,69 +0,0 @@
-// Type definitions for core-decorators.js 0.19
```
no longer need this because TypeScript is generating individual `.d.ts` files per source file.

BurtHarris commented on index.ts:
```diff
@@ -1,4 +0,0 @@
-// index.ts - facade for core-decorators
```
Replaced with a change to `main` in `package.json`

Commit:
insert_final_newline

Commit:
Whitespace fixes

BurtHarris commented on .editorconfig:
```diff
@@ -10,10 +10,14 @@ end_of_line = lf
 indent_style = space
+charset = utf-8
+end_of_line = lf
+insert_final_newline = true
```
I'm not sure why, but insert_final_newline this isn't working.   Probably a problem on my side...

BurtHarris replied:
Ahh, I had the editorconfig add-in for vscode installed but disabled...   That explains some things.

Commit:
Removed unneeded change to .editorconfig

Commit:
+ comment re: yarn's upcoming 1.0 release

Commit:
Clarifying yarn comment in gulpfile

jayphelps commented on src/core-decorators.ts:
```diff
@@ -8,20 +8,19 @@
 export { default as profile } from './profile';
 
+// Exported for testing purposes
+export  {default as defaultConsole} from './defaultConsole';
```
please indent like the others.

```js
export { default as defaultConsole } from './defaultConsole';
```

BurtHarris replied:
Fixed

BurtHarris replied:
P.S.   I'd like to add tslint too, currently these escaped eslint because of the file extension.

jayphelps replied:
Groovy!

jayphelps commented on src/deprecate.ts:
```diff
@@ -22,6 +24,10 @@ function handleDescriptor(target, key, descriptor, [msg = DEFAULT_MSG, options =
   return decorate(handleDescriptor, args);
 }
+
+export {deprecate, deprecate as deprecated} ;
```
indention

BurtHarris replied:
Fixed

jayphelps commented on src/lazy-initialize.ts:
```diff
@@ -30,6 +30,9 @@ function handleDescriptor(target, key, descriptor) {
   return decorate(handleDescriptor, args);
 }
+
+export {lazyInitialize};
```
indention

BurtHarris replied:
Fixed.

jayphelps commented on src/private/utils.ts:
```diff
@@ -30,19 +30,7 @@ export function decorate(handleDescriptor, entryArgs) {
-  throttleTrailingArgs = null;
-
-  @lazyInitialize
+  // @lazyInitialize
```
I think this is extraneous?

BurtHarris replied:
I remember commenting out the @lazyInitialize part was intentional.   Its broken due to TypeScript's limitations on property initializers, tests from Babel fail with the @lazyInitialize in place.

As I remember removing the `throttleTrailingArgs = null;` part was part of removing the obsolete decorators and their tests.

jayphelps replied:
sorry if I'm misunderstanding but it's still not clear if this comment extraneous? It seems like it is

BurtHarris replied:
Oh!   You mean I could simply delete the comment.   Of course you are right.

jayphelps commented on src/profile.ts:
```diff
@@ -1,15 +1,9 @@
-import { decorate, metaFor, warn, bind } from './private/utils';
+import { decorate, metaFor } from './private/utils';
+import {defaultConsole} from './defaultConsole';
```
indention

BurtHarris replied:
fixed

jayphelps commented on test/exports.spec.js:
```diff
@@ -1,15 +1,14 @@
+const glob = require('glob');
+const camelCase = require('camelCase');
+const interopRequire = require('interop-require');
+const decorators = require('core-decorators');
```
Why did we lose the ability to use ESM imports? sorry if I missed a previous explanation.

BurtHarris replied:
What I was shooting for was that the exports.spec.js and test.spec.js files should not require any transpiler.   I think as long as we're running tests on a modern version of node.js, we could use ESM imports here, but I wasn't sure.   I can restore them if you want.   Please advise.

jayphelps replied:
@BurtHarris I wasn't going to burden you _right now_ with it, but I was hoping to also switch the tests over to TypeScript so that we're actually also testing the type definitions at the same time. This can come later (if ever). For now your plan is OK with me.

BurtHarris replied:
@jayphelps I had shared your hope, just want to make sure you understand that as of TypeScript@2.4, the limited support for property decorators means we can't really test _all_ of core-decorators from a TypeScript file.    This is the quote from the [TypeScript handbook ](https://www.typescriptlang.org/docs/handbook/decorators.html) I think reflects the limitation:

> NOTE  A Property Descriptor is not provided as an argument to a property decorator due to how property decorators are initialized in TypeScript. This is because there is currently no mechanism to describe an instance property when defining members of a prototype, and no way to observe or modify the initializer for a property. As such, a property decorator can only be used to observe that a property of a specific name has been declared for a class.

I've tried to find issues tracking this restriction, Microsoft/TypeScript#11866 seems one, but it's been closed, making reference to possibly reopening once the [Public class fields TC39 proposal](https://tc39.github.io/proposal-class-public-fields/) is ratified.   I don't fully understand that proposal, but it seems different from the stage 3 decorators one.   Can you see any reason they couldn't/shouldn't address it the way Babel does, by passing a synthetic (transpiler generated) descriptor to the first decorator in the list?

jayphelps replied:
Oh yes..I forgot they didn't support it. I say we punt.

Re TS updating to stage-3: I haven't heard, but I would bet money TS is waiting for the decorators proposal to be finalized stage-4 because in its current stage-3 form it's going to be quite the pain in the ass of a breaking change for the TS community aka mostly Angular people, so they want to be 100% sure of the changes so they only have to do it once.

BurtHarris replied:
But doesn't getting to stage 4 require "Two compatible implementations which pass the acceptance tests"?   I was thinking TypeScript & Babel might be the likely implementations, have I misread the TC39 process and practices?

jayphelps commented on test/unit/override.spec.js:
```diff
@@ -36,7 +36,7 @@ describe('@override', function () {
       class Child extends Parent {
         @override
-        speak(first, second) {}
+        speak (first, second) {}
```
hmm why add the space in-between all of these? I'd prefer they weren't.

BurtHarris replied:
Personally, I agree with you.   The requirement for a space here is part of the eslint "semistandard" ruleset, and I'm not strongly advocating it....    As I said somewhere else, I picked that because it seemed a close match to existing code, but this whitespace was one change needed to conform to "semistandard".      

I think picking a canned standard is easier than maintaining our own ruleset, but I don't have broad knowledge of all the rulesets out there.

jayphelps reviewed (changes requested):
It's a pretty big PR so it'll take me a bit to really dive deep into it and confirm, but I did an initial pass and semantically everything appeared great. I only found code style consistency issues.

Commit:
Whitespace/indentation fixes

BurtHarris commented:
Yes, sorry for the size of the PR.   

I've fixed the indentation issues.  They slipped by because __eslint__ doesn't understand typescript files, the long-term fix is to add __tslint__, but I didn't want to bloat the PR even more.   I'll add a workitem to track that.

On the space after function name issue, I agree with you.   The requirement for a space there is part of the eslint "semistandard" ruleset.     As I said somewhere else, I picked that ruleset because it seemed a close match to existing code, but this whitespace was one change needed to conform to "semistandard".      

I think picking a canned standard is easier than maintaining our own ruleset, but I don't have broad knowledge of all the rulesets out there.  If you want to choose a different ruleset, or have me tweak that rule I'm happy to oblige.

jayphelps commented:
> Yes, sorry for the size of the PR @jayphelps.

No need to apologize, by its nature it has to be huge--but I appreciate it.

> If you want to choose a different ruleset, or have me tweak that rule I'm happy to oblige.

I don't have a preference on standardized ruleset vs custom either way; I've always maintained my own, and IMO it's not a huge deal. I do however personally have a problem with the space in the CallExpressions, that we're talking about. If you'd prefer it, we can of course discuss it further, but if it's all the same to you let's not have the space.

BurtHarris commented:
On the topic of code style, I think that the "recommended" style that comes with eslint may be a better match for both of our sensibilities.   I'll try it out and update the branch.    

P.S.  I wish there were a "do-not-merge" label I could attach to reflect the fact I still plan on touchups.   Is there a way for me to add such a label?

Commit:
Removed extranious comment in util.ts

jayphelps commented:
You should have permissions to create new labels and then add them to any tickets, if not lmk. That won't truly prevent you or I from merging it though. I've changed the settings now to disabling merge ability until at least one review approval. We'll see how it works.

Commit:
eslint tune-up

BurtHarris commented:
@jayphelps I've updated eslint to extend the "reccomended" set and added a rule for the function-parens spacing, fixing the violations.   There are 15 lint warnings remaining, these are issues I'd like your input on at some time, particularly the `no-unused-expressions` ones.

BurtHarris commented:
Current test status on my machine:
- Babel tests: 0 failures
- TypeScript tests: 10 failures

```
[12:28:03] Using gulpfile C:\code\core-decorators.js\gulpfile.js
[12:28:03] Starting 'symlink'...
[12:28:03] Finished 'symlink' after 1.08 ms
[12:28:03] Starting 'test'...
[12:28:03] Finished 'test' after 11 ms

  Main package exports
    √ exports all defined decorator files inside core-decorators.js

  applyDecorators() helper (from Babel)
    √ applies the decorators to the provided prop's descriptors

  @autobind (from Babel)
    √ returns a bound instance for a method
    √ sets the correct prototype descriptor options
    √ sets the correct instance descriptor options when bound
    √ sets the correct instance descriptor options when reassigned outside
    √ works with multiple instances of the same class
    √ returns the same bound function every time
    √ works with inheritance, super.method() being autobound as well
    √ throws when it needs WeakMap but it is not available
    √ does not override descriptor when accessed on the prototype
    √ can be used to autobind the entire class at once
    √ correctly binds with multiple class prototype levels
    √ correctly binds class with symbol properties

  @decorate (from Babel)
    √ correctly applies user provided function to method
    √ sets the correct prototype descriptor options
    √ is tied to the instance, not the prototype

  @deprecate (from Babel)
DEPRECATION Foo#first: This function will be removed in future versions.
DEPRECATION Foo#first: This function will be removed in future versions.
    √ console.warn() is called with default warning when the deprecated function is used
    √ console.warn() is called with the custom message, when provided
DEPRECATION Foo#forth: fdsa

    See http://example.com/ for more details.


    √ console.warn() is called with the URL, when provided

  @enumerable (from Babel)
    √ is enumerable

  @lazyInitialize (from Babel)
    √ does not initialize property until it the getter is called
    √ allows normal reassignment
    √ does not initialize property when looked up on the prototype directly

  @nonconfigurable (from Babel)
    √ is marked configurable: false
    √ causes `delete` to be noop
    √ throws if you try to reconfigure it

  @nonenumerable (from Babel)
    √ is not enumerable

  @override (from Babel)
    √ throws error when signature does not match
    √ throws error when no is matching name is found
    √ throws error when no is matching name is found but suggests a closely named method if exists
    √ does not throw an error when signatures match

  @profile (from Babel)
    √ calls console.profile and console.profileEnd
    √ calls console.profile and console.profileEnd once when flag is on
    √ calls console.profileEnd even if the called method throws
    √ uses class and method names for a default prefix
    √ uses a supplied prefix for the label
    √ supports a custom profileation object
    √ returns the value
    when throttled (from Babel)
      √ should always call function
      √ should call profile after throttle time
    when functioned (from Babel)
      √ should have `this` context
      √ should accept parameters
    when disabled (from Babel)
      √ should send a warning
      √ should leave descriptor alone

  @readonly (from Babel)
    √ marks descriptor as writable === false
    √ makes setting property error

  @suppressWarnings (from Babel)
    √ suppresses warns
    √ restores the original console.warn

  @time (from Babel)
    √ calls console.time and console.timeEnd
    √ calls console.timeEnd even if the called method throws
    √ uses class and method names for a default prefix
    √ creates a unique label with a counter
    √ uses a supplied prefix for the label
    √ supports a custom timeation object
    √ returns the value

  applyDecorators() helper (from TypeScript)
    √ applies the decorators to the provided prop's descriptors

  @autobind (from TypeScript)
    √ returns a bound instance for a method
    1) sets the correct prototype descriptor options
    √ sets the correct instance descriptor options when bound
    √ sets the correct instance descriptor options when reassigned outside
    √ works with multiple instances of the same class
    √ returns the same bound function every time
    2) works with inheritance, super.method() being autobound as well
    3) throws when it needs WeakMap but it is not available
    √ does not override descriptor when accessed on the prototype
    √ can be used to autobind the entire class at once
    √ correctly binds with multiple class prototype levels
    √ correctly binds class with symbol properties

  @decorate (from TypeScript)
    √ correctly applies user provided function to method
    4) sets the correct prototype descriptor options
    √ is tied to the instance, not the prototype

  @deprecate (from TypeScript)
    5) "before each" hook for "console.warn() is called with default warning when the deprecated function is used"
    6) "after each" hook for "console.warn() is called with default warning when the deprecated function is used"

  @enumerable (from TypeScript)
    √ is enumerable

  @lazyInitialize (from TypeScript)
    7) does not initialize property until it the getter is called
    8) allows normal reassignment
    √ does not initialize property when looked up on the prototype directly

  @nonconfigurable (from TypeScript)
    √ is marked configurable: false
    √ causes `delete` to be noop
    √ throws if you try to reconfigure it

  @nonenumerable (from TypeScript)
    9) is not enumerable

  @override (from TypeScript)
    √ throws error when signature does not match
    √ throws error when no is matching name is found
    √ throws error when no is matching name is found but suggests a closely named method if exists
    √ does not throw an error when signatures match

  @profile (from TypeScript)
    √ calls console.profile and console.profileEnd
    √ calls console.profile and console.profileEnd once when flag is on
    √ calls console.profileEnd even if the called method throws
    √ uses class and method names for a default prefix
    √ uses a supplied prefix for the label
    √ supports a custom profileation object
    √ returns the value
    when throttled (from TypeScript)
      √ should always call function
      √ should call profile after throttle time
    when functioned (from TypeScript)
      √ should have `this` context
      √ should accept parameters
    when disabled (from TypeScript)
      √ should send a warning
      √ should leave descriptor alone

  @readonly (from TypeScript)
    √ marks descriptor as writable === false
    10) makes setting property error

  @suppressWarnings (from TypeScript)
    √ suppresses warns
    √ restores the original console.warn

  @time (from TypeScript)
    √ calls console.time and console.timeEnd
    √ calls console.timeEnd even if the called method throws
    √ uses class and method names for a default prefix
    √ creates a unique label with a counter
    √ uses a supplied prefix for the label
    √ supports a custom timeation object
    √ returns the value


  100 passing (126ms)
  10 failing

  1) @autobind (from TypeScript) sets the correct prototype descriptor options:

      enumerable property mismatch
      + expected - actual

      -true
      +false

      at Context.<anonymous> (test\typescript\autobind.spec.ts:75:28)

  2) @autobind (from TypeScript) works with inheritance, super.method() being autobound as well:
     TypeError: Cannot read property 'should' of undefined
      at Context.<anonymous> (test\typescript\autobind.spec.ts:143:28)

  3) @autobind (from TypeScript) throws when it needs WeakMap but it is not available:
     AssertionError: expected [Function] to throw an error
      at Context.<anonymous> (test\typescript\autobind.spec.ts:160:20)

  4) @decorate (from TypeScript) sets the correct prototype descriptor options:

      enumerable
      + expected - actual

      -true
      +false

      at Context.<anonymous> (test\typescript\decorate.spec.ts:55:28)

  5) @deprecate (from TypeScript) "before each" hook for "console.warn() is called with default warning when the deprecated function is used":
     TypeError: Cannot read property 'spy' of undefined
      at Context.<anonymous> (test\typescript\deprecate.spec.ts:28:10)

  6) @deprecate (from TypeScript) "after each" hook for "console.warn() is called with default warning when the deprecated function is used":
     TypeError: utils.warn.restore is not a function
      at Context.<anonymous> (test\typescript\deprecate.spec.ts:32:16)

  7) @lazyInitialize (from TypeScript) does not initialize property until it the getter is called:
     TypeError: Cannot assign to read only property 'bar' of object '#<Foo>'
      at new Foo (test\typescript\lazy-initialize.spec.ts:10:8)
      at Context.<anonymous> (test\typescript\lazy-initialize.spec.ts:22:17)

  8) @lazyInitialize (from TypeScript) allows normal reassignment:
     TypeError: Cannot assign to read only property 'bar' of object '#<Foo>'
      at new Foo (test\typescript\lazy-initialize.spec.ts:10:8)
      at Context.<anonymous> (test\typescript\lazy-initialize.spec.ts:30:17)

  9) @nonenumerable (from TypeScript) is not enumerable:
     TypeError: Cannot assign to read only property 'bar' of object '#<Foo>'
      at new Foo (test\typescript\nonenumerable.spec.ts:6:8)
      at Context.<anonymous> (test\typescript\nonenumerable.spec.ts:10:17)

  10) @readonly (from TypeScript) makes setting property error:
     TypeError: Cannot assign to read only property 'second' of object '#<Foo>'
      at new Foo (test\typescript\readonly.spec.ts:9:11)
      at Context.<anonymous> (test\typescript\readonly.spec.ts:18:17)



(node:21320) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 2): Error in plugin 'gulp-mocha'
Message:
    Command failed: mocha C:\code\core-decorators.js\test\babel\applyDecorators.spec.js C:\code\core-decorators.js\test\babel\autobind.spec.js C:\code\core-decorators.js\test\babel\decorate.spec.js C:\code\core-decorators.js\test\babel\deprecate.spec.js C:\code\core-decorators.js\test\babel\enumerable.spec.js C:\code\core-decorators.js\test\babel\lazy-initialize.spec.js C:\code\core-decorators.js\test\babel\nonconfigurable.spec.js C:\code\core-decorators.js\test\babel\nonenumerable.spec.js C:\code\core-decorators.js\test\babel\override.spec.js C:\code\core-decorators.js\test\babel\profile.spec.js C:\code\core-decorators.js\test\babel\readonly.spec.js C:\code\core-decorators.js\test\babel\suppress-warnings.spec.js C:\code\core-decorators.js\test\babel\time.spec.js C:\code\core-decorators.js\test\typescript\applyDecorators.spec.js C:\code\core-decorators.js\test\typescript\autobind.spec.js C:\code\core-decorators.js\test\typescript\decorate.spec.js C:\code\core-decorators.js\test\typescript\deprecate.spec.js C:\code\core-decorators.js\test\typescript\enumerable.spec.js C:\code\core-decorators.js\test\typescript\lazy-initialize.spec.js C:\code\core-decorators.js\test\typescript\nonconfigurable.spec.js C:\code\core-decorators.js\test\typescript\nonenumerable.spec.js C:\code\core-decorators.js\test\typescript\override.spec.js C:\code\core-decorators.js\test\typescript\profile.spec.js C:\code\core-decorators.js\test\typescript\readonly.spec.js C:\code\core-decorators.js\test\typescript\suppress-warnings.spec.js C:\code\core-decorators.js\test\typescript\time.spec.js --colors
DEPRECATION Foo#first: This function will be removed in future versions.
DEPRECATION Foo#first: This function will be removed in future versions.
DEPRECATION Foo#forth: fdsa

    See http://example.com/ for more details.




  Main package exports
    √ exports all defined decorator files inside core-decorators.js

  applyDecorators() helper (from Babel)
    √ applies the decorators to the provided prop's descriptors

  @autobind (from Babel)
    √ returns a bound instance for a method
    √ sets the correct prototype descriptor options
    √ sets the correct instance descriptor options when bound
    √ sets the correct instance descriptor options when reassigned outside
    √ works with multiple instances of the same class
    √ returns the same bound function every time
    √ works with inheritance, super.method() being autobound as well
    √ throws when it needs WeakMap but it is not available
    √ does not override descriptor when accessed on the prototype
    √ can be used to autobind the entire class at once
    √ correctly binds with multiple class prototype levels
    √ correctly binds class with symbol properties

  @decorate (from Babel)
    √ correctly applies user provided function to method
    √ sets the correct prototype descriptor options
    √ is tied to the instance, not the prototype

  @deprecate (from Babel)
    √ console.warn() is called with default warning when the deprecated function is used
    √ console.warn() is called with the custom message, when provided
    √ console.warn() is called with the URL, when provided

  @enumerable (from Babel)
    √ is enumerable

  @lazyInitialize (from Babel)
    √ does not initialize property until it the getter is called
    √ allows normal reassignment
    √ does not initialize property when looked up on the prototype directly

  @nonconfigurable (from Babel)
    √ is marked configurable: false
    √ causes `delete` to be noop
    √ throws if you try to reconfigure it

  @nonenumerable (from Babel)
    √ is not enumerable

  @override (from Babel)
    √ throws error when signature does not match
    √ throws error when no is matching name is found
    √ throws error when no is matching name is found but suggests a closely named method if exists
    √ does not throw an error when signatures match

  @profile (from Babel)
    √ calls console.profile and console.profileEnd
    √ calls console.profile and console.profileEnd once when flag is on
    √ calls console.profileEnd even if the called method throws
    √ uses class and method names for a default prefix
    √ uses a supplied prefix for the label
    √ supports a custom profileation object
    √ returns the value
    when throttled (from Babel)
      √ should always call function
      √ should call profile after throttle time
    when functioned (from Babel)
      √ should have `this` context
      √ should accept parameters
    when disabled (from Babel)
      √ should send a warning
      √ should leave descriptor alone

  @readonly (from Babel)
    √ marks descriptor as writable === false
    √ makes setting property error

  @suppressWarnings (from Babel)
    √ suppresses warns
    √ restores the original console.warn

  @time (from Babel)
    √ calls console.time and console.timeEnd
    √ calls console.timeEnd even if the called method throws
    √ uses class and method names for a default prefix
    √ creates a unique label with a counter
    √ uses a supplied prefix for the label
    √ supports a custom timeation object
    √ returns the value

  applyDecorators() helper (from TypeScript)
    √ applies the decorators to the provided prop's descriptors

  @autobind (from TypeScript)
    √ returns a bound instance for a method
    1) sets the correct prototype descriptor options
    √ sets the correct instance descriptor options when bound
    √ sets the correct instance descriptor options when reassigned outside
    √ works with multiple instances of the same class
    √ returns the same bound function every time
    2) works with inheritance, super.method() being autobound as well
    3) throws when it needs WeakMap but it is not available
    √ does not override descriptor when accessed on the prototype
    √ can be used to autobind the entire class at once
    √ correctly binds with multiple class prototype levels
    √ correctly binds class with symbol properties

  @decorate (from TypeScript)
    √ correctly applies user provided function to method
    4) sets the correct prototype descriptor options
    √ is tied to the instance, not the prototype

  @deprecate (from TypeScript)
    5) "before each" hook for "console.warn() is called with default warning when the deprecated function is used"
    6) "after each" hook for "console.warn() is called with default warning when the deprecated function is used"

  @enumerable (from TypeScript)
    √ is enumerable

  @lazyInitialize (from TypeScript)
    7) does not initialize property until it the getter is called
    8) allows normal reassignment
    √ does not initialize property when looked up on the prototype directly

  @nonconfigurable (from TypeScript)
    √ is marked configurable: false
    √ causes `delete` to be noop
    √ throws if you try to reconfigure it

  @nonenumerable (from TypeScript)
    9) is not enumerable

  @override (from TypeScript)
    √ throws error when signature does not match
    √ throws error when no is matching name is found
    √ throws error when no is matching name is found but suggests a closely named method if exists
    √ does not throw an error when signatures match

  @profile (from TypeScript)
    √ calls console.profile and console.profileEnd
    √ calls console.profile and console.profileEnd once when flag is on
    √ calls console.profileEnd even if the called method throws
    √ uses class and method names for a default prefix
    √ uses a supplied prefix for the label
    √ supports a custom profileation object
    √ returns the value
    when throttled (from TypeScript)
      √ should always call function
      √ should call profile after throttle time
    when functioned (from TypeScript)
      √ should have `this` context
      √ should accept parameters
    when disabled (from TypeScript)
      √ should send a warning
      √ should leave descriptor alone

  @readonly (from TypeScript)
    √ marks descriptor as writable === false
    10) makes setting property error

  @suppressWarnings (from TypeScript)
    √ suppresses warns
    √ restores the original console.warn

  @time (from TypeScript)
    √ calls console.time and console.timeEnd
    √ calls console.timeEnd even if the called method throws
    √ uses class and method names for a default prefix
    √ creates a unique label with a counter
    √ uses a supplied prefix for the label
    √ supports a custom timeation object
    √ returns the value


  100 passing (126ms)
  10 failing

  1) @autobind (from TypeScript) sets the correct prototype descriptor options:

      enumerable property mismatch
      + expected - actual

      -true
      +false

      at Context.<anonymous> (test\typescript\autobind.spec.ts:75:28)

  2) @autobind (from TypeScript) works with inheritance, super.method() being autobound as well:
     TypeError: Cannot read property 'should' of undefined
      at Context.<anonymous> (test\typescript\autobind.spec.ts:143:28)

  3) @autobind (from TypeScript) throws when it needs WeakMap but it is not available:
     AssertionError: expected [Function] to throw an error
      at Context.<anonymous> (test\typescript\autobind.spec.ts:160:20)

  4) @decorate (from TypeScript) sets the correct prototype descriptor options:

      enumerable
      + expected - actual

      -true
      +false

      at Context.<anonymous> (test\typescript\decorate.spec.ts:55:28)

  5) @deprecate (from TypeScript) "before each" hook for "console.warn() is called with default warning when the deprecated function is used":
     TypeError: Cannot read property 'spy' of undefined
      at Context.<anonymous> (test\typescript\deprecate.spec.ts:28:10)

  6) @deprecate (from TypeScript) "after each" hook for "console.warn() is called with default warning when the deprecated function is used":
     TypeError: utils.warn.restore is not a function
      at Context.<anonymous> (test\typescript\deprecate.spec.ts:32:16)

  7) @lazyInitialize (from TypeScript) does not initialize property until it the getter is called:
     TypeError: Cannot assign to read only property 'bar' of object '#<Foo>'
      at new Foo (test\typescript\lazy-initialize.spec.ts:10:8)
      at Context.<anonymous> (test\typescript\lazy-initialize.spec.ts:22:17)

  8) @lazyInitialize (from TypeScript) allows normal reassignment:
     TypeError: Cannot assign to read only property 'bar' of object '#<Foo>'
      at new Foo (test\typescript\lazy-initialize.spec.ts:10:8)
      at Context.<anonymous> (test\typescript\lazy-initialize.spec.ts:30:17)

  9) @nonenumerable (from TypeScript) is not enumerable:
     TypeError: Cannot assign to read only property 'bar' of object '#<Foo>'
      at new Foo (test\typescript\nonenumerable.spec.ts:6:8)
      at Context.<anonymous> (test\typescript\nonenumerable.spec.ts:10:17)

  10) @readonly (from TypeScript) makes setting property error:
     TypeError: Cannot assign to read only property 'second' of object '#<Foo>'
      at new Foo (test\typescript\readonly.spec.ts:9:11)
      at Context.<anonymous> (test\typescript\readonly.spec.ts:18:17)
```

BurtHarris commented:
@jayphelps My tests match up well with [Appveyor](https://ci.appveyor.com/project/jayphelps/core-decorators-js/build/1.0.25), but [Travis.CI](https://travis-ci.org/jayphelps/core-decorators.js/builds/261126996?utm_source=github_status&utm_medium=notification) seems to have other problems.    I'm uncomfortable that neither of these were considered a failure and that the PR claims that "All checks have passed."   Thoughts?

jayphelps commented:
Travis is reporting success because `gulp test` returned exit status 0, indicating success by POSIX convention. I'm not immediately familiar with gulp enough to know how to bubble up a non-zero exit status e.g. 1

jayphelps commented on test/exports.spec.js:
```diff
@@ -1,32 +1,31 @@
+const chai = require('chai');
+const path = require('path');
+const glob = require('glob');
+const camelCase = require('camelCase');
```
This is what is causing the error on Linux. This import string needs to be `camelcase` all lowercase. Linux filesystem is case-sensitive.

Commit:
Support Yarn

Due to a complicated bug, npm is acting strange for me.   I've found the Yarn package manager doesn't have the same prolem.   Added a yarn.lock file, and a note in README.md.

Having both a yarn and npm lock file seems like it might be undesirable long term, but we can reexamine this once some one of the npm/VSCode/typescript bugs are resolved.

Commit:
cleanup remaining lint & test issues

Some tests have been marked so that they are skipped in typescript, but these show up as pending rather than failed.

These include:
 - Two tests in autobind.spec.js (needs review)
 - Two in lazyIntialize.spec.json
 - One in readonly.spec.json
 - One in nonenmerable.spec.json

Commit:
Update describing TypeScript support in README.md

Commit:
fix module capitilization

This matters on Linux.

BurtHarris commented:
@jayphelps,

- I've fixed that test error exit status problem by eliminating the `gulp-mocha` interop layer. just launch `mocha` to run the full set if tests 
- I've fixed all the code style issues
- I've fix all tests running, with six exceptions.   These have been marked as pending for the TypeScript tests, so they don't block successful completion of the test suite.   I think four of the remaining problem tests have something to do with the TypeScript property decorator restriction.    Two other, in __autobind.spec.js__ I'm less sure about, I think they may be detecting something else.   In general I think autobind is working.   
- I've made an update to the README to reflect the level of TypeScript support and mention the restrictions above.   
- I think it's close to ready.

jayphelps commented:
Unfortunately we may have jumped the gun here. Babel is actively working on [adding support for stage-2 decorators](https://github.com/babel/proposals/issues/11) and when they do, I think core-decorators needs to follow suit. AFAIK TypeScript has not yet offered any public indication of when (or even if) they're going to do so as well. Thoughts?

jayphelps commented:
One thing we could do, that I don't _like_ (but this situation sucks no matter what) is indeed convert core-decorators to TypeScript, release it and continue to maintain it as normal then when babel gets stage-2+ spec we go back to babel but maintain the TypeScript version in a branch. We could continue to make bug fixes on it (bumping the semver patch on that same locked minor version).

jayphelps commented:
Just had another thought: I haven't looked into deeply yet, but it may be possible for us to detect which spec version is being used by the user simply by the arguments passed in. Duck-type checking or instanceof. If that's the case, we could still in fact write this in TypeScript and have it check at runtime which one is being used. The trouble is we need to test the Babel codepath, and I don't think it's possible for TypeScript to compile to something that would work in that case--but we might be able to get away with only using syntax extensions that are also supported by [babel's Flow preset](https://www.npmjs.com/package/babel-preset-flow).

BurtHarris commented:
Reviewing TypeScript PRs scheduled for inclusion in the next version I found Microsoft/TypeScript#19675, which looks like it may address the remaining issue blocking this.    If I get some cycles I'll try building this with TypeScript@next and verify.
